packages/polling-controller/src/AbstractPollingController.ts

```diff
@@ -0,0 +1,138 @@
+import type { NetworkClientId } from '@metamask/network-controller';
+import type { Json } from '@metamask/utils';
+import stringify from 'fast-json-stable-stringify';
+import { v4 as random } from 'uuid';
+
+export type IPollingController = {
+  startPollingByNetworkClientId(
+    networkClientId: NetworkClientId,
+    options: Json,
+  ): string;
+
+  stopAllPolling(): void;
+
+  stopPollingByPollingToken(pollingToken: string): void;
+
+  onPollingCompleteByNetworkClientId(
+    networkClientId: NetworkClientId,
+    callback: (networkClientId: NetworkClientId) => void,
+    options: Json,
+  ): void;
+
+  _executePoll(networkClientId: NetworkClientId, options: Json): Promise<void>;
+  _startPollingByNetworkClientId(
+    networkClientId: NetworkClientId,
+    options: Json,
+  ): void;
+  _stopPollingByPollingTokenSetId(key: PollingTokenSetId): void;
+};
+
+export const getKey = (
+  networkClientId: NetworkClientId,
+  options: Json,
+): PollingTokenSetId => `${networkClientId}:${stringify(options)}`;
+
+export type PollingTokenSetId = `${NetworkClientId}:${string}`;
+
+type Constructor = new (...args: any[]) => object;
+
+/**
+ * AbstractPollingControllerBaseMixin
+ *
+ * @param Base - The base class to mix onto.
+ * @returns The composed class.
+ */
+export function AbstractPollingControllerBaseMixin<TBase extends Constructor>(
+  Base: TBase,
+) {
+  abstract class AbstractPollingControllerBase
+    extends Base
+    implements IPollingController
+  {
+    readonly #pollingTokenSets: Map<PollingTokenSetId, Set<string>> = new Map();
+
+    #callbacks: Map<
+      PollingTokenSetId,
+      Set<(PollingTokenSetId: PollingTokenSetId) => void>
+    > = new Map();
+
+    abstract _executePoll(
+      networkClientId: NetworkClientId,
+      options: Json,
+    ): Promise<void>;
+
+    abstract _startPollingByNetworkClientId(
+      networkClientId: NetworkClientId,
+      options: Json,
+    ): void;
+
+    abstract _stopPollingByPollingTokenSetId(key: PollingTokenSetId): void;
+
+    startPollingByNetworkClientId(
+      networkClientId: NetworkClientId,
+      options: Json = {},
+    ): string {
+      const pollToken = random();
+      const key = getKey(networkClientId, options);
+      const pollingTokenSet =
+        this.#pollingTokenSets.get(key) || new Set<string>();
```

Comment:
Nit: `??` would be safer:

``` suggestion
        this.#pollingTokenSets.get(key) ?? new Set<string>();

Reply:
done here: https://github.com/MetaMask/core/pull/3636/commits/823aaa188f572a8aa76973a012ae9bfa8da518e0